// yuejia/component/Calendar/Day.tsx
import yj from '../../typings';
import * as React from 'react';
import * as classNames from 'classnames';
import './index.scss';

export interface BaseProps {
  onClick: (e: React.MouseEvent<HTMLDivElement>, day: number) => void;
}

export interface Props extends yj.ComponentProps, BaseProps {
  /** 显示的文本 */
  label: number;
  /** 是否显示出来， */
  visible: boolean;
  /** 是否激活 */
  active: boolean;
  /** 是否在label下面加点 */
  dot: boolean;
  /** 是否是周末 */
  weekend: boolean;
  /** 是否是选中的日期 */
  selected: boolean;
}

const prefixCls = 'yj-component-calendar-day';

const Day = (props: Props) => {
  const {
    className, label, visible, active, dot, weekend, onClick, selected
  } = props;
  return (
    <div
      className={classNames(prefixCls, className, {
        [`${prefixCls}-active`]: active,
        [`${prefixCls}-weekend`]: weekend,
        [`${prefixCls}-selected`]: selected
      })}
      style={{
        visibility: visible ? 'visible' : 'hidden'
      }}
      onClick={(e) => onClick(e, label)}
    >
      <div className={`${prefixCls}-label`}>{label}</div>
      {dot ? (
        <div className={`${prefixCls}-dot`}>&bull;</div>
      ) : null}
    </div>
  );
};

export default Day;
