// yuejia/component/Carousel/Slide.tsx
import yj from '../../typings';
import * as React from 'react';
import * as classNames from 'classnames';
import './index.scss';

export interface BaseProps {
  /** 是否开启循环 */
  loop?: boolean;
  children?: JSX.Element[] | JSX.Element;
  /** 整体slide的高度 */
  slideHeight?: number;
}

interface Props extends yj.ComponentProps, BaseProps {
  /** 单个slide宽度 */
  slideWidth: number;
  /** slide总个数 */
  slideCount: number;
}

const prefixCls = 'yj-component-carousel';

// 设置容器总宽度
function setFrameWidth(slideWidth: number, count: number, loop?: boolean): number {
  let width = 0;

  for (let i = 0; i < count; i++) {
    width += slideWidth;
  }

  if (loop) {
    width += 2 * slideWidth;
  }

  return width;
}

const Slide = (props: Props) => {
  const { className, children, slideWidth, slideHeight, style, loop, slideCount } = props;
  // slide总宽度
  const frameWidth = setFrameWidth(slideWidth, slideCount, loop);

  return (
    <div
      className={classNames(`${prefixCls}-slide`, className)}
      style={{
        width: frameWidth,
        height: slideHeight,
        ...style
      }}
    >
      {children && React.Children.map(children, (child, i) => {
        return React.cloneElement(child as React.ReactElement<any>, {
          className: classNames(`${prefixCls}-slide-item`, (child as React.ReactElement<any>).props.className),
          style: {
            width: slideWidth,
            ...(child as React.ReactElement<any>).props.style
          }
        });
      })}
    </div>
  );
};

export default Slide;
