// yuejia/component/Map/Control/Zoom.tsx
import * as React from 'react';
import * as className from 'classnames';
import Icon from '../../Icon';
import Container from '../../Container';
import svgRootPath from '../../../utils/svgRootPath';
import sheet, { getActionName } from '../../../model/behavior/sheet';
import PageStack from '../../PageStack';
import { ChildrenProps } from '../index';
import './index.scss';
const svgFile = {
  minus: svgRootPath + require('../../../assets/minus.svg').id,
  plus: svgRootPath + require('../../../assets/plus.svg').id,
};

interface Props extends ChildrenProps {
}

const prefixCls = 'yj-component-map-control';
class Zoom extends React.PureComponent<Props> {
  constructor(props: Props) {
    super(props);

    this.onClick = this.onClick;
  }

  private onClick(type: string): void {
    const { map } = this.props;
    const stackComponent = PageStack.getStackComponent();

    if (map) {
      if (type === 'minus') {
        map.zoomOut();

        if (stackComponent) {
          stackComponent.behavior.trigger({
            actionName: getActionName(sheet.MapControl, 0, this.props),
          });
        }
      } else {
        map.zoomIn();

        if (stackComponent) {
          stackComponent.behavior.trigger({
            actionName: getActionName(sheet.MapControl, 1, this.props),
          });
        }
      }
    }
  }

  public render(): JSX.Element[] {
    return ['minus', 'plus'].map((item) => (
      <Container
        className={`${prefixCls}-item`}
        alignItems="center"
        justifyContent="center"
        key={item}
        onClick={() => this.onClick(item)}
      >
        <Icon
          className={className(`${prefixCls}-svg`, `${prefixCls}-zoom`)}
          src={svgFile[item as 'minus' | 'plus']}
        />
      </Container>
    ));
  }
}

export default Zoom;
