// yuejia/component/UploadImg/Img.tsx
import * as React from 'react';
import * as classNames from 'classnames';
import Icon from '../Icon';
import ImgView from '../ImgView';
import svgRootPath from '../../utils/svgRootPath';
const closeSvg = svgRootPath + require('../../assets/cross-circle-o.svg').id;

export interface Props {
  src: string;
  failView?: string;
  index: number;
  count: number;
  onClick: (src: string) => void;
  onFailClick: () => void;
  onRemove: (index: number) => void;
  disabled: boolean;
  itemCls?: string;
}

const prefixCls = 'yj-component-uploadimg';
const Img = (props: Props) => {
  const { src, failView, onClick, index, onRemove, disabled, itemCls, count, onFailClick } = props;
  return (
    <div className={classNames(itemCls, {
      [`${prefixCls}-item`]: count > 1,
      [`${prefixCls}-single`]: count === 1,
    })}>
      <ImgView
        className={`${prefixCls}-img`}
        // loading={imgFile.headPhotoLoading}
        completeView={src}
        failView={failView}
        onClick={() => onClick(src)}
        onFailClick={onFailClick}
      />
      {!disabled ? (
        <Icon
          className={`${prefixCls}-remove`}
          src={closeSvg}
          onClick={() => onRemove(index)}
        />
      ) : null}
    </div>
  );
};

export default Img;
