// yuejia/component/UploadImg/Upload.tsx
import * as React from 'react';
import * as classNames from 'classnames';
import Icon from '../Icon';
import isSvgFile from '../../utils/isSvgFile';
import Container from '../Container';

interface Base {
  onClick: () => void;
}

export interface Props {
  uploadIcon: string;
  count: number;
  uploadIconCls?: string;
  uploadCls?: string;
  itemCls?: string;
}

const prefixCls = 'yj-component-uploadimg';

const Upload = (props: Props & Base) => {
  const { uploadIcon, uploadIconCls, uploadCls, onClick, itemCls, count } = props;
  return (
    <Container
      className={classNames(`${prefixCls}-upload`, uploadCls, itemCls, {
        [`${prefixCls}-item`]: count > 1,
        [`${prefixCls}-single`]: count === 1,
      })}
      alignItems="center"
      justifyContent="center"
      onClick={onClick}
    >
      {isSvgFile(uploadIcon) ? (
        <Icon className={uploadIconCls} src={uploadIcon} />
      ) : (
        <img className={uploadIconCls} src={uploadIcon} />
      )}
    </Container>
  );
};

export default Upload;
