// yuejia/model/behavior/sheet.ts
import yj from '../../typings';

export interface DetailParams {
  Banner: {
    /** 图片的id */
    id: string | undefined;
  };
  Calendar: {
    /** 选择的日期 */
    date: string;
  };
  Filter: {
    /** 值 */
    value: string;
    /** 文本 */
    text: string;
  };
  ImgView: {
    /** 图片的id */
    id: string | undefined;
  };
  Location: {
    /** 定位到的省 */
    province: string;
    /** 定位到的城市 */
    city: string;
    /** 定位到的地区 */
    district: string;
    /** 经度 */
    lng: number;
    /** 纬度 */
    lat: number;
  };
  Search: {
    /** 搜索内容的关键字 */
    keyword: string;
  };
  Tab: {
    /** 被点击的名称 */
    name: string;
  };
  Tel: {
    /** 拨打的电话 */
    tel: string;
  };
  MapNavigation: {
    /** 导航要去的目的地 */
    destination: string;
  };
  MapMarkerBehavior: {
    /** 经度 */
    lng: number;
    /** 纬度 */
    lat: number;
  };
  Video: {
    /** 视频的id */
    id: string | undefined;
  };
}

interface Option {
  /** 动作，一个动词 */
  action: string[];
  /** 组件名称，用于跟动作组合 */
  componentName: string;
}

interface Sheet {
  [key: string]: Option;
}

export function getActionName(option: Option, actionIndex: number, props: yj.ComponentBehavior): string {
  const componentName = props.componentName || option.componentName;

  return componentName ? `${option.action[actionIndex]}-${componentName}` : `${option.action[actionIndex]}`;
}

const sheet = {
  Button: {
    action: ['点击'],
    componentName: '按钮'
  },
  Banner: {
    action: ['滑动', '点击'],
    componentName: '轮播图'
  },
  Calendar: {
    action: ['点击'],
    componentName: '日历'
  },
  Collection: {
    action: ['添加', '取消'],
    componentName: '收藏'
  },
  Filter: {
    action: ['筛选'],
    componentName: ''
  },
  ImgView: {
    action: ['点击'],
    componentName: '图片'
  },
  Location: {
    action: ['定位'],
    componentName: '位置'
  },
  Search: {
    action: ['搜索'],
    componentName: '搜索'
  },
  Popup: {
    action: ['点击'],
    componentName: '弹出层'
  },
  Tab: {
    action: ['点击'],
    componentName: '选项卡'
  },
  Tel: {
    action: ['拨打'],
    componentName: '电话'
  },
  MapControl: {
    action: ['点击-缩小', '点击-放大', '点击-回到中心点'],
    componentName: ''
  },
  MapNavigation: {
    action: ['点击'],
    componentName: '导航'
  },
  MapMarkerBehavior: {
    action: ['点击'],
    componentName: '标注'
  },
  Video: {
    action: ['点击'],
    componentName: '视频'
  }
};

export default sheet as Sheet;
