// yuejia/model/location/index.ts
import Map from '../../component/Map/service';
import GeolocationService, { Address } from '../../component/Map/service/geolocation';
import { ReadOnly } from '../../_base/datatype';

interface Props {
  /** 默认定位 */
  location: Address;
  /** 默认选中省市 */
  address?: Address;
  /** 储存坐标的aoi */
  saveApi: string;
  defaultLocation: Address;
}

class Location {
  constructor(props: Props) {
    this.map = new Map();
    this.map.geolocation = new GeolocationService({
      failAddress: props.location,
      saveApi: props.saveApi,
      value: props.address
    });
    this.defaultLocation = props.defaultLocation;
    this.getValue = this.getValue.bind(this);
    this.setValue = this.setValue.bind(this);
  }

  public map: Map;
  private defaultLocation: Address;

  /** 优先选择的省市， */
  public getValue(): ReadOnly<Address> {
    const address = this.map.geolocation.address.get();
    const defaultLocation = this.defaultLocation;
    return {
      province: address ? address.province : defaultLocation.province,
      city: address ? address.city : defaultLocation.city,
      cityId: address ? address.cityId : defaultLocation.cityId,
    };
  }

  public setValue(address: Address): void {
    this.map.geolocation.address.set(address);
  }
}

export default Location;
