// yuejia-cli/lib/behavior/createXlsx/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const fs = require("fs");
const XLSX = require("xlsx-style");
function createXlsx(data) {
    const pkg = fs.readFileSync('./package.json', { encoding: 'utf8' });
    const pkgObj = JSON.parse(pkg);
    const version = pkgObj.version;
    const project = pkgObj.description;
    const workSheet = {};
    const keyMap = ['页面名称', '行为名称', '详细参数'];
    const colsCount = keyMap.length;
    const rows = [keyMap, ...data.map((item) => [item.pageName, item.actionName, item.detailParams])];
    const rowsCount = rows.length;
    const range = {
        s: { c: 0, r: 0 },
        e: { c: colsCount - 1, r: rowsCount - 1 }
    };
    const merges = [];
    let startRow = 0;
    let endRow = 0;
    workSheet['!ref'] = XLSX.utils.encode_range(range);
    for (let rowIndex = 0; rowIndex < rowsCount; rowIndex++) {
        if (rows[rowIndex][0] !== rows[endRow][0]) {
            merges.push({
                s: { c: 0, r: startRow },
                e: { c: 0, r: endRow },
            });
            startRow = rowIndex;
        }
        endRow = rowIndex;
        for (let colIndex = 0; colIndex < colsCount; colIndex++) {
            const cell_ref = XLSX.utils.encode_cell({ c: colIndex, r: rowIndex });
            const excelCell = {
                v: rows[rowIndex][colIndex],
                t: 's',
                s: {
                    alignment: {
                        wrapText: true,
                        vertical: 'center'
                    }
                }
            };
            workSheet[cell_ref] = excelCell;
        }
    }
    workSheet['!cols'] = [{
            wpx: 300
        }, {
            wpx: 300
        }, {
            wpx: 300
        }];
    workSheet['!merges'] = merges;
    XLSX.writeFile({
        Sheets: {
            默认: workSheet
        },
        SheetNames: ['默认'],
    }, `${project}${version}行为数据.xlsx`, {
        bookType: 'xlsx',
        type: 'file',
    });
    console.log(`${project}${version}行为数据.xlsx 生成完毕`.green);
}
exports.default = createXlsx;
