// yuejia-cli/lib/create/dir.ts
import * as fs from 'fs';
import * as path from 'path';
import createFile from './file';

interface FileParams {
  writeFileName: string;
  writeContent: string;
  readFilePath?: string;
}

export default function createDir(dirpath: string, fileParams: FileParams[]): void {
  if (!fs.existsSync(dirpath)) {
    let pathtmp: string;
    dirpath.split(path.sep).forEach((dirname: string) => {
      pathtmp = pathtmp ? path.join(pathtmp, dirname) : dirname;

      if (!fs.existsSync(pathtmp)) {
        if (!fs.mkdirSync(pathtmp)) {
          for (let i = 0; i < fileParams.length; i++) {
            if (fileParams[i].writeFileName) {
              createFile({
                filepath: pathtmp,
                filename: fileParams[i].writeFileName,
                writeContent: fileParams[i].writeContent
              }, fileParams[i].readFilePath);
            }
          }
          return true;
        }
      }
    });
  }
}

/** 根据文件路径创建目录，会排除文件名 */
export function createDirByFilePath(dirpath: string): void {
  const split = dirpath.split(path.sep);
  let finalPath = './';

  for (let i = 0; i < split.length - 1; i++) {
    finalPath += split[i] + '/';
    createDir(finalPath, []);
  }
}
