// yuejia/component/Map/service/index.ts
import Geolocation from './geolocation';
import Marker from './marker';
import Supporting from './supporting';
import ObObject from '../../../app/datatype/obObject';

export type MapMarker = Map<string, Marker[]>;

class MapModel {
  /** 百度地图对象 */
  public map: BMap.Map | undefined;
  /** 定位 */
  public geolocation: Geolocation = new Geolocation();
  /** 周边配套 */
  public supporting: Supporting = new Supporting();
  /** 周边配套标注  */
  public supportingMarker: ObObject<MapMarker> = new ObObject({ defaultValue: new Map() });
  /** 中心点 */
  public center: Marker = new Marker();
  /** 激活的标注 */
  public activeMarker: ObObject<Marker | null> = new ObObject({ defaultValue: this.center });

  /** 初始化 */
  public init(map: BMap.Map): void {
    this.map = map;
  }

  /** 从定位点到地图中心点的距离，单位米，要在定位成功之后调用 */
  public getDistanceByCenter(): number {
    if (this.map) {
      const geolocation = this.geolocation.getValue();
      const end = this.center.mapPoint;

      if (geolocation && end) {
        return this.map.getDistance(geolocation.point as BMap.Point, end);
      }
    }

    return 0;
  }
}

export default MapModel;
