// yuejia/component/Banner/Carousel.tsx
import * as React from 'react';
import Carousel, { BaseProps as CarouselComponentProps } from '../Carousel';
import { BaseProps as ImgViewProps } from '../ImgView';
import { BaseProps as StateBaseProps } from '../StateMachine';
import Item, { Data, Props as ItemProps } from './Item';
import ObArray from '../../app/datatype/obArray';

export interface CarouselBaseProps extends CarouselComponentProps, StateBaseProps, ImgViewProps {
  /** 单个slide的样式 */
  itemCls?: string;
  /** 单个图片的样式 */
  imgCls?: string;
  /** 空视图 */
  emptyView?: any;
  /** 失败的视图 */
  failView?: string;
  /** 视频封面 */
  poster?: ItemProps['poster'];
}

export interface CarouselProps extends CarouselBaseProps  {
  index: number;
  /** 渲染的数据 */
  data: ObArray<Data>;
  /** 点击事件，会把当前的数据传入 */
  onClick: (data: Data) => void;
  /** 用于判断是否是手动指定跳转 */
  goToPageId: any;
}

const AppCarousel = (props: CarouselProps) => {
  const {
    dots, autoplay, onAfterSlide, shadow, index, loadingStyle, mode, failView,
    slideHeight, itemCls, imgCls, onClick, goToPageId, swiping, poster
  } = props;
  const data = props.data.get();
  const dataId = props.data.getDataId();

  return (
    <Carousel
      loop
      autoplay={autoplay && data.length > 1}
      interval={3000}
      swiping={swiping ? data.length > 1 : false}
      dots={dots && data.length > 1}
      onAfterSlide={onAfterSlide}
      index={index}
      slideHeight={slideHeight}
      dataId={dataId}
      goToPageId={goToPageId}
    >
      <Item
        data={data}
        shadow={shadow}
        loadingStyle={loadingStyle}
        mode={mode}
        failView={failView}
        itemCls={itemCls}
        imgCls={imgCls}
        dataId={dataId}
        onClick={onClick}
        poster={poster || data[0].src}
      />
    </Carousel>
  );
};

export default AppCarousel;
