// yuejia/component/Container/index.tsx
import yj from '../../typings';
import * as React from 'react';
import * as classNames from 'classnames';
import Border, { Props as BorderProps } from '../Border';
import './index.scss';

export interface Props extends yj.ComponentProps {
  children?: any | any[];
  retinaBorder?: BorderProps;
  flexDirection?: 'row' | 'row-reverse' | 'column' | 'column-reverse';
  flexWrap?: 'nowrap' | 'wrap' | 'wrap-reverse';
  justifyContent?: 'flex-start' | 'flex-end' | 'center' | 'space-between' | 'space-around';
  alignItems?: 'flex-start' | 'flex-end' | 'center' | 'baseline' | 'stretch';
  alignContent?: 'flex-start' | 'flex-end' | 'center' | 'space-between' | 'space-around' | 'stretch';
  order?: number;
  flex?: number;
  alignSelf?: 'auto' | 'flex-start' | 'flex-end' | 'center' | 'baseline' | 'stretch';
  onClick?: yj.OnClick<HTMLDivElement>;
  /** 是启用touchumove事件 */
  touchumove?: boolean;
}

export const prefixCls = 'yj-component-container';
const Container = (props: Props) => {
  const {
    className, style, flexDirection, flexWrap, justifyContent, alignItems, alignContent, children, order, id,
    flex, alignSelf, onClick, retinaBorder
  } = props;

  return (
    <div
      className={classNames(prefixCls, className, {
        [`${prefixCls}-flex`]: !!(justifyContent || alignItems || alignSelf || flexDirection),
        [`${prefixCls}-flexDirection-${flexDirection}`]: !!flexDirection,
        [`${prefixCls}-flexWrap-${flexWrap}`]: !!flexWrap,
        [`${prefixCls}-justifyContent-${justifyContent}`]: !!justifyContent,
        [`${prefixCls}-alignItems-${alignItems}`]: !!alignItems,
        [`${prefixCls}-alignContent-${alignContent}`]: !!alignContent,
        [`${prefixCls}-order-${order}`]: !!order,
        [`${prefixCls}-flex-${flex}`]: !!flex,
        [`${prefixCls}-alignSelf-${alignSelf}`]: !!alignSelf,
      })}
      style={style}
      id={id}
      onClick={onClick}
    >
      {children}
      {retinaBorder ? (
        <Border
          direction={retinaBorder.direction}
          color={retinaBorder.color}
        />
      ) : null}
    </div>
  );
};

export default Container;
